plugins/network-elements/nuage-vsp/src/com/cloud/network/manager/NuageVspManagerImpl.java

```diff
@@ -285,60 +507,341 @@ public boolean deleteNuageVspDevice(DeleteNuageVspDeviceCmd cmd) {
+            public AgentControlAnswer processControlCommand(long agentId, AgentControlCommand cmd) {
+                return null;
+            }
+
+            @Override
+            public void processConnect(Host host, StartupCommand cmd, boolean forRebalance) throws ConnectionException {
+
+            }
+
+            @Override
+            public boolean processDisconnect(long agentId, Status state) {
+                return true;
+            }
+
+            @Override
+            public boolean isRecurring() {
+                return false;
+            }
+
+            @Override
+            public int getTimeout() {
+                return 0;
+            }
+
+            @Override
+            public boolean processTimeout(long agentId, long seq) {
+                return true;
+            }
+        }, false, true, false);
+    }
+
+    @DB
+    private void initNuageNetworkOffering() {
+        Transaction.execute(new TransactionCallbackNoReturn() {
+            @Override
+            public void doInTransactionWithoutResult(TransactionStatus status) {
+                if (_networkOfferingDao.findByUniqueName(nuageVspSharedNetworkOfferingWithSGServiceName) == null) {
+                    NetworkOfferingVO defaultNuageVspSharedSGNetworkOffering =
+                            new NetworkOfferingVO(nuageVspSharedNetworkOfferingWithSGServiceName, "Offering for NuageVsp Shared Security group enabled networks",
+                                    Networks.TrafficType.Guest, false, false, null, null, true, NetworkOffering.Availability.Optional, null, Network.GuestType.Shared, true, true, false, false, false);
+
+                    defaultNuageVspSharedSGNetworkOffering.setState(NetworkOffering.State.Enabled);
+                    defaultNuageVspSharedSGNetworkOffering = _networkOfferingDao.persistDefaultNetworkOffering(defaultNuageVspSharedSGNetworkOffering);
+
+                    Map<Network.Service, Network.Provider> defaultNuageVspSharedSGNetworkOfferingProviders = new HashMap<Network.Service, Network.Provider>();
+                    defaultNuageVspSharedSGNetworkOfferingProviders.put(Network.Service.Dhcp, Network.Provider.NuageVsp);
+                    defaultNuageVspSharedSGNetworkOfferingProviders.put(Network.Service.SecurityGroup, Network.Provider.NuageVsp);
+                    defaultNuageVspSharedSGNetworkOfferingProviders.put(Network.Service.Connectivity, Network.Provider.NuageVsp);
+
+                    for (Network.Service service : defaultNuageVspSharedSGNetworkOfferingProviders.keySet()) {
+                        NetworkOfferingServiceMapVO offService =
+                                new NetworkOfferingServiceMapVO(defaultNuageVspSharedSGNetworkOffering.getId(), service, defaultNuageVspSharedSGNetworkOfferingProviders.get(service));
+                        _networkOfferingServiceMapDao.persist(offService);
+                        s_logger.trace("Added service for the NuageVsp network offering: " + offService);
```

Comment:
Wrap the call to ``s_logger.trace`` in a ``if (s_logger.isTraceEnabled()) { }`` to avoid pressure on the string pool when trace logging is disabled.